### ISSUE TITLE
set_keepalive reports EINVAL on Linux

### ISSUE THREAD

**Issue by zonyitoo** [alexcrichton/net2-rs#13]
Linux is Ubuntu with 3.13.0-63-generic kernel, x86_64.

Error debug message is

```
Error { repr: Os { code: 22, message: "Invalid argument" } }
```

The same code runs well on OS X 10.10.5.

Call from the `mio::tcp::TcpStream::set_keepalive` with argument `Some(300)`.

**Comment by alexcrichton** [comment]
Perhaps the socket is in an invalid state? Do you have some example code I could try out?

**Comment by zonyitoo** [comment]
Just call `set_keepalive` immediately after you accepted a new `TcpStream`.

**Comment by zonyitoo** [comment]
Modified from MIO's official example

``` rust
extern crate mio;

use mio::*;
use mio::tcp::{TcpListener, TcpStream};

// Setup some tokens to allow us to identify which event is
// for which socket.
const SERVER: Token = Token(0);
const CLIENT: Token = Token(1);

fn main() {
    let addr = "127.0.0.1:3000".parse().unwrap();

    // Setup the server socket
    let server = TcpListener::bind(&addr).unwrap();

    // Create an event loop
    let mut event_loop = EventLoop::new().unwrap();

    // Start listening for incoming connections
    event_loop.register(&server, SERVER, EventSet::readable(),
                        PollOpt::edge()).unwrap();

    // Setup the client socket
    let sock = TcpStream::connect(&addr).unwrap();

    // Register the socket
    event_loop.register(&sock, CLIENT, EventSet::readable(),
                        PollOpt::edge()).unwrap();

    // Define a handler to process the events
    struct MyHandler(TcpListener);

    impl Handler for MyHandler {
        type Timeout = ();
        type Message = ();

        fn ready(&mut self, event_loop: &mut EventLoop<MyHandler>, token: Token, _: EventSet) {
            match token {
                SERVER => {
                    let MyHandler(ref mut server) = *self;

                    println!("Got Readable event from TcpListener");
                    match server.accept() {
                        Ok(None) => {},
                        Ok(Some((stream, addr))) => {
                            println!("Got connection from {:?}", addr);
                            // Call set_keepalive
                            if let Err(err) = stream.set_keepalive(Some(300)) {
                                println!("Error occurs while set_keepalive: {:?}", err);
                            }
                        },
                        Err(err) => {
                            panic!("Error occurs while accepting: {:?}", err);
                        }
                    }
                }
                CLIENT => {
                    // The server just shuts down the socket, let's just
                    // shutdown the event loop
                    event_loop.shutdown();
                }
                _ => panic!("unexpected token"),
            }
        }
    }

    // Start handling events
    event_loop.run(&mut MyHandler(server)).unwrap();
}
```

The output is

```
Got Readable event from TcpListener
Got connection from V4(127.0.0.1:39653)
Error occurs while set_keepalive: Error { repr: Os { code: 22, message: "Invalid argument" } }
```

**Comment by alexcrichton** [comment]
I believe the error here is that `set_keepalive` in mio is specified in seconds, where `set_keepalive_ms` on net2 is specified in milliseconds. Using a keepalive timeout of 300s apparently is too high for linux? Going down to 300ms (is that the intention?) works, however.

Regardless though I think everything's working as expected in this library, it may just be surprising behavior coming out of linux!

**Comment by carllerche** [comment]
The man page describes `EINVAL` as

> optlen invalid in setsockopt(). In some cases this error can also occur for an invalid value in optval (e.g., for the IP_ADD_MEMBERSHIP option described in ip(7)).

**Comment by zonyitoo** [comment]
Yes, I want to set 300s!

**Comment by zonyitoo** [comment]
Actually, even I modified `Some(300)` to `Some(60)`, it still got `EINVAL`.

**Comment by zonyitoo** [comment]
According to this line of code: https://github.com/alexcrichton/net2-rs/blob/master/src/ext.rs#L590

In `man tcp`, it said

```
       TCP_KEEPIDLE (since Linux 2.4)
              The time (in seconds) the connection needs to remain idle before TCP starts sending  keepalive  probes,
              if  the socket option SO_KEEPALIVE has been set on this socket.  This option should not be used in code
              intended to be portable.
```

So the argument of `TCP_KEEPIDLE` should be seconds instead of milliseconds.

I think this issue should be re-opened.

**Comment by alexcrichton** [comment]
Aha, thanks for the sharp eye! I updated the impl here and published 0.2.15, could you try that out?

**Comment by zonyitoo** [comment]
@alexcrichton Well, it works fine. Thanks.
